examples/Tutorial 4 - 3SAT Solver.ipynb

```diff
@@ -0,0 +1,477 @@
+    "        return True\n",
+    "        \n",
+    "    \n",
+    "\n",
+    "    def __len__(self) -> int:\n",
+    "        \"\"\"\n",
+    "        The number of clauses in this instance.\n",
+    "        \"\"\"\n",
+    "        return len(self.clauses)\n",
+    "    \n",
+    "\n",
+    "    def compile_to_mis(self) -> MISInstance:\n",
+    "        \"\"\"\n",
+    "            Compile this instance of 3SAT into an instance of MIS.\n",
+    "        \"\"\"\n",
+    "        \n",
+    "        \"\"\"\n",
+    "        Making a dictionary such that the keys are of the form \"XF\" and \"XT\", such that nodes[\"XT\"] is the list of\n",
+    "        nodes representing the term (\"X\", True), and nodes[\"XT\"] is the list of nodes representing the term (\"X\", False) \n",
+    "        \"\"\"\n",
+    "        nodes = defaultdict(lambda : []) # set the default value of a key to an empty list\n",
+    "        \n",
+    "        get_key = lambda term, reverse : term.variable + (\"F\" if term.positive ^ reverse else \"T\")\n",
+    "        \n",
+    "        graph = nx.Graph()\n",
+    "        for i in range(len(self.clauses)):\n",
+    "            graph.add_edges_from([(3*i,3*i+1), (3*i+1, 3*i+2), (3*i, 3*i+2)])\n",
+    "            # add edges between each node and its complementaries\n",
+    "            for j in range(3):\n",
+    "                for node in nodes[get_key(self.clauses[i].terms[j], True)]:\n",
+    "                    graph.add_edge(3*i + j, node)\n",
+    "            \n",
+    "            # add the new nodes to the cache\n",
+    "            for j in range(3):\n",
+    "                nodes[get_key(self.clauses[i].terms[j], False)].append(3*i+j)\n",
+    "                        \n",
+    "        return MISInstance(graph)\n",
+    "            \n",
+    "    \n",
+    "\n",
+    "    def rebuild_result_from_mis(self, solutions: list[MISSolution]) -> dict[str, bool] | None:\n",
+    "        \"\"\"\n",
+    "        Search in a list of MISSolution if any of them is a possible solution to 3SAT.\n",
+    "        \"\"\"\n",
+    "        if solutions is None:\n",
+    "            return None\n",
+    "        \n",
+    "        for solution in solutions:\n",
+    "            if len(solution.nodes) != len(self.clauses): \n",
+    "                continue # skip if the number of nodes isn't enough to be a solution \n",
+    "            # otherwise, we have a solution\n",
+    "            result = dict()\n",
+    "            for variable in self._variables:\n",
+    "                result[variable] = False # set the default of every variable to False in case the current solution makes them free\n",
+    "                \n",
+    "            for node in solution.nodes:\n",
+    "                term = self.clauses[node//3].terms[node%3] # since each node has index equal to 3*clause + term\n",
+    "                result[term.variable] = term.positive\n",
+    "            \n",
```

Comment:
Since the quantum algorithm is non-deterministic, it may sometimes return MIS solutions that are false positives.

Let's check that `self.eval(result)` is `True` before returning, to make sure we don't return a false 3SAT solution (and `continue` otherwise).